### PR TITLE
Stable9.1 ldap event logging

### PR DESCRIPTION
**Pull request by butonic**
## Description
This PR adds a LoggingEventLogger that can be enabled by setting `'debug' => true, 'log.events' => true,` in config.php (other log_* config options should be migrated to the . notation, but that is subject of a different PR) 
It also adds event logging for ldap read and search events.

@PVince81 @jvillafanez Don't cry when you see what the app actually does ...

## Related Issue

https://github.com/owncloud/core/pull/27083

## Motivation and Context
ldap performance problems

## How Has This Been Tested?
local instance with openldap

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by butonic** [comment]
lets say there is room for improvement. First we do an ldap search, then for every result we do ... multiple reads:
```console
$ sudo -u www-data php /var/www/oc/stable9.1/occ ldap:search zombie400 --limit 1
Otte, Courtney (zombie400) (8dce7e96-d23b-1035-9670-9bb74d28f7e6)
```
```json
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"start ldap_search (LDAP search baseDN:[\"dc=example,dc=org\"] filter:\"(&(|(objectclass=posixAccount))(displayName=*)(|(uid=zombie400*)(mail=zombie400*)(displayName=zombie400*)))\" attr:[\"dn\",\"uid\",\"samaccountname\",\"memberof\",\"\",\"\",\"displayName\",\"uid\"] attrOnly:0 limit:0)","level":0,"time":"2017-02-09T14:53:28+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"end ldap_search (LDAP search baseDN:[\"dc=example,dc=org\"] filter:\"(&(|(objectclass=posixAccount))(displayName=*)(|(uid=zombie400*)(mail=zombie400*)(displayName=zombie400*)))\" attr:[\"dn\",\"uid\",\"samaccountname\",\"memberof\",\"\",\"\",\"displayName\",\"uid\"] attrOnly:0 limit:0), took 0.134045s","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"start ldap_read (LDAP read baseDN:\"uid=zombie400,ou=zombies,dc=example,dc=org\" filter:\"(|(objectclass=posixAccount))\" attr:[\"\"])","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"end ldap_read (LDAP read baseDN:\"uid=zombie400,ou=zombies,dc=example,dc=org\" filter:\"(|(objectclass=posixAccount))\" attr:[\"\"]), took 0.002194s","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"start ldap_read (LDAP read baseDN:\"uid=zombie400,ou=zombies,dc=example,dc=org\" filter:\"objectClass=*\" attr:[\"uid\"])","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"end ldap_read (LDAP read baseDN:\"uid=zombie400,ou=zombies,dc=example,dc=org\" filter:\"objectClass=*\" attr:[\"uid\"]), took 0.001216s","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"start ldap_read (LDAP read baseDN:\"uid=zombie400,ou=zombies,dc=example,dc=org\" filter:\"objectClass=*\" attr:[\"displayName\"])","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"ExDJ+s2pA\/WdCpRZDwsg","remoteAddr":"","app":"diagnostics","message":"end ldap_read (LDAP read baseDN:\"uid=zombie400,ou=zombies,dc=example,dc=org\" filter:\"objectClass=*\" attr:[\"displayName\"]), took 0.001221s","level":0,"time":"2017-02-09T14:53:29+00:00","method":"--","url":"--","user":"--"}
{"reqId":"zSwx9\/rVwhDJ6twtTZ+4","remoteAddr":"","app":"diagnostics","message":"start ldap_search (LDAP search baseDN:[\"dc=example,dc=org\"] filter:\"(&(|(objectclass=posixAccount))(displayName=*)(|(uid=zombie400*)(mail=zombie400*)(displayName=zombie400*)))\" attr:[\"dn\",\"uid\",\"samaccountname\",\"memberof\",\"\",\"\",\"displayName\",\"uid\"] attrOnly:0 limit:0)","level":0,"time":"2017-02-09T14:53:34+00:00","method":"--","url":"--","user":"--"}
{"reqId":"zSwx9\/rVwhDJ6twtTZ+4","remoteAddr":"","app":"diagnostics","message":"end ldap_search (LDAP search baseDN:[\"dc=example,dc=org\"] filter:\"(&(|(objectclass=posixAccount))(displayName=*)(|(uid=zombie400*)(mail=zombie400*)(displayName=zombie400*)))\" attr:[\"dn\",\"uid\",\"samaccountname\",\"memberof\",\"\",\"\",\"displayName\",\"uid\"] attrOnly:0 limit:0), took 0.047777s","level":0,"time":"2017-02-09T14:53:34+00:00","method":"--","url":"--","user":"--"}
```

File: apps/user_ldap/lib/LDAP.php [hunk 1]
```diff
@@ -175,7 +175,10 @@ public function nextEntry($link, $result) {
 	 * @return mixed
 	 */
 	public function read($link, $baseDN, $filter, $attr) {
-		return $this->invokeLDAPMethod('read', $link, $baseDN, $filter, $attr);
+		\OC::$server->getEventLogger()->start('ldap_read', "LDAP read baseDN:".json_encode($baseDN).' filter:'.json_encode($filter).' attr:'.json_encode($attr));
```
Comment (PVince81) [inline apps/user_ldap/lib/LDAP.php]: I guess `$link` is not needed ? (no idea what that is...)

File: lib/private/Security/CSRF/CsrfTokenManager.php [hunk 2]
```diff
@@ -44,6 +44,10 @@ public function __construct(CsrfTokenGenerator $tokenGenerator,
 		$this->sessionStorage = $storageInterface;
 	}
 
+	public function setSessionStorage(SessionStorage $sessionStorage) {
+		$this->sessionStorage = $sessionStorage;
```
Comment (PVince81) [inline lib/private/Security/CSRF/CsrfTokenManager.php]: uh, why is this needed ?

**Review by PVince81 — COMMENTED** [review]
Code looks fine, but I don't get the part about session storage

File: apps/user_ldap/lib/LDAP.php (hunk 1 above)
Comment (butonic) [inline apps/user_ldap/lib/LDAP.php]: link is the ldap connection

File: lib/private/Security/CSRF/CsrfTokenManager.php (hunk 2 above)
Comment (butonic) [inline lib/private/Security/CSRF/CsrfTokenManager.php]: oh well ... since I am using the Diagnostics Event Logger stuff the initial event logging in base.php initializes the CsrfTokenManager with a SessionStorage that is based on the in memory session. Then we update the session in https://github.com/owncloud/core/blob/master/lib/base.php#L428 which closes the in memory session. Then when we actually want to generate the csrf token the in memory session is closed, leading to a Session Already Closed Exception ... 

In any case, this allows using logging early on ...

Comment (butonic) [inline lib/private/Security/CSRF/CsrfTokenManager.php]: If there is a cleaner way, I am happy for enlightenment.

Comment (PVince81) [inline lib/private/Security/CSRF/CsrfTokenManager.php]: Ah ok. I didn't see it used anywhere in the logging code so didn't see how it fits together.

Comment (butonic) [inline lib/private/Security/CSRF/CsrfTokenManager.php]: If you comment https://github.com/owncloud/core/pull/27125#discussion-diff-100324708R48 you will see the exception right after pressing f5 in the browser ... ;)

**Comment by PVince81** [comment]
Tested after patching 9.1.3 EE, seems to work fine.

**Comment by jvillafanez** [comment]
This might need a twist... if several apps are using this, I should be able to choose what app I want to debug somehow. Otherwise we end up with tons of log about apps that we're not interested into.

And the options to enable this should be documented somewhere.

**Comment by PVince81** [comment]
keep or close ?

**Comment by DeepDiver1975** [comment]
9.1 will be eol this month
